internal/provider/resource_application.go

```diff
@@ -122,6 +129,92 @@ func (r *applicationResource) Configure(ctx context.Context, req resource.Config
 	r.subCtx = tflog.NewSubsystem(ctx, LogResourceApplication)
 }
 
+func storageSetRequiresReplace(ctx context.Context, req planmodifier.SetRequest, resp *setplanmodifier.RequiresReplaceIfFuncResponse) {
+
+	planSet := make(map[string]jujustorage.Constraints)
+	if !req.PlanValue.IsNull() {
+		var planStorageSlice []nestedStorage
+		resp.Diagnostics.Append(req.PlanValue.ElementsAs(ctx, &planStorageSlice, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		if len(planStorageSlice) > 0 {
+			for _, storage := range planStorageSlice {
+				storageName := storage.Label.ValueString()
+				storageSize := storage.Size.ValueString()
+				storagePool := storage.Pool.ValueString()
+				storageCount := storage.Count.ValueInt64()
+
+				// Validate storage size
+				parsedStorageSize, err := utils.ParseSize(storageSize)
+				if err != nil {
+					resp.Diagnostics.AddError("Invalid StorageContraints Size", fmt.Sprintf("Invalid storage size %q: %s", storageSize, err))
+					return
+				}
+
+				planSet[storageName] = jujustorage.Constraints{
+					Size:  parsedStorageSize,
+					Pool:  storagePool,
+					Count: uint64(storageCount),
+				}
+			}
+		}
+	}
+
+	stateSet := make(map[string]jujustorage.Constraints)
+	if !req.StateValue.IsNull() {
```

Comment:
Investigate whether you need to check req.StateValue.IsUnknown() also

Reply:
it does not need to happen because we use this function in Update.